deeppavlov/evolve.py

```diff
@@ -0,0 +1,322 @@
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.commands.utils import set_deeppavlov_root, expand_path
+
+log = get_logger(__name__)
+
+parser = argparse.ArgumentParser()
+
+parser.add_argument("config_path", help="path to a pipeline json config", type=str)
+parser.add_argument('--key_main_model', help='key inserted in dictionary of main model in pipe', default="main")
+parser.add_argument('--p_cross', help='probability of crossover', type=float, default=0.2)
+parser.add_argument('--pow_cross', help='crossover power', type=float, default=0.1)
+parser.add_argument('--p_mut', help='probability of mutation', type=float, default=1.)
+parser.add_argument('--pow_mut', help='mutation power', type=float, default=0.1)
+
+parser.add_argument('--p_size', help='population size', type=int, default=10)
+parser.add_argument('--gpus', help='visible GPUs divided by comma <<,>>', default="-1")
+parser.add_argument('--train_partition',
+                    help='partition of splitted train file', default=1)
+parser.add_argument('--start_from_population',
+                    help='population number to start from. 0 means from scratch', default=0)
+parser.add_argument('--path_to_population',
+                    help='path to population to start from', default="")
+parser.add_argument('--elitism_with_weights',
+                    help='whether to save elite models with weights or without', default=0)
+parser.add_argument('--iterations', help='Number of iterations', type=int, default=-1)
+
+
+def find_config(pipeline_config_path: str):
+    if not Path(pipeline_config_path).is_file():
+        configs = [c for c in Path(__file__).parent.glob(f'configs/**/{pipeline_config_path}.json')
+                   if str(c.with_suffix('')).endswith(pipeline_config_path)]  # a simple way to not allow * and ?
+        if configs:
+            log.info(f"Interpreting '{pipeline_config_path}' as '{configs[0]}'")
+            pipeline_config_path = str(configs[0])
+    return pipeline_config_path
+
+
+def main():
+    args = parser.parse_args()
+
+    pipeline_config_path = find_config(args.config_path)
+    key_main_model = args.key_main_model
+    population_size = args.p_size
+    gpus = [int(gpu) for gpu in args.gpus.split(",")]
+    train_partition = int(args.train_partition)
+    start_from_population = int(args.start_from_population)
+    path_to_population = args.path_to_population
+    elitism_with_weights = int(args.elitism_with_weights)
+    iterations = int(args.iterations)
+
+    p_crossover = args.p_cross
+    pow_crossover = args.pow_cross
+    p_mutation = args.p_mut
+    pow_mutation = args.pow_mut
+
+    if os.environ.get("CUDA_VISIBLE_DEVICES") is None:
+        pass
+    else:
+        cvd = [int(gpu) for gpu in os.environ.get("CUDA_VISIBLE_DEVICES").split(",")]
+        if set(gpus).issubset(set(cvd)):
+            pass
+        else:
+            try:
+                gpus = [cvd[gpu] for gpu in gpus]
```

Comment:
I think that `gpus = [cvd[gpu] for gpu in gpus]` should be used even `if set(gpus).issubset(set(cvd))`

Reply:
if set(gpus).issubset(set(cvd)), user means use 3,5 out of 3,5,7 visible devices.
else user means use 0,1 out of 3,5,6 visible devices. 
Used devices will be the same but the way of assigning is different.

Reply:
for the second comment, it will raise error for this case because it is really not clear what the user wants

Reply:
the second is fixed. it will raise error because it is unclear what to use